### DISCUSSION TITLE
formatUrl

### DISCUSSION THREAD

**Discussion by Amerehei** [vercel/next.js#54540] [App Router]
Is there any way to create a url in `string`? I found that there is an internal `formatUrl` method. is there any public method?

https://github.com/vercel/next.js/blob/7ac6bb1a88395953a0dc2706269be696e98ac460/packages/next/src/shared/lib/router/utils/format-url.ts#L29

**Comment by devjiwonchoi** [comment]
Could you share more context of which case you need help? For instance, route, page, or just a URL object to string?

**Comment by Amerehei** [comment]
I need to convert a `UrlObject` to string. Let me explain more, I need a method to be aware of Next.JS config like `basePath`

**Reply by devjiwonchoi** [reply]
Just wondering if `toString()` isn't working for your case.

**Reply by Amerehei** [reply]
what about `basePath`? how can I include it to the url? I don't want re-invent the wheel
see https://github.com/vercel/next.js/blob/canary/packages/next/src/client/add-base-path.ts#L6

**Reply by devjiwonchoi** [reply]
Got it. In route handlers like `middleware | route.ts`, use `NextRequest.nextUrl.basePath`.

```ts
// route.ts

import { type NextRequest, NextResponse } from 'next/server'

export async function GET(request: NextRequest) {
  const { basePath } = request.nextUrl
  return NextResponse.json({ basePath })
}
```

@Amerehei Also note that `basePath` is supported in various cases like redirect, rewrite, headers, Link, etc.
So it'd be good for you to consider if you do need to actually retrieve the basePath.

### Reference

- https://nextjs.org/docs/app/api-reference/next-config-js/headers#headers-with-basepath-support
- https://nextjs.org/docs/app/api-reference/next-config-js/basePath#links
- https://nextjs.org/docs/app/api-reference/next-config-js/headers#headers-with-basepath-support
- https://nextjs.org/docs/app/api-reference/next-config-js/redirects#redirects-with-basepath-support
- https://nextjs.org/docs/app/api-reference/next-config-js/rewrites#rewrites-with-basepath-support
